### PR TITLE
Fix gateway workspace header forwarding and expand MCP run/artifact tools

### LINKED ISSUE #23001: [BUG] Gateway adapter does not forward workspace context when invoking judge endpoints
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

## Issues Policy acknowledgement

I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

## Where did you encounter this bug?

Local machine (self-hosted MLflow with workspaces enabled)

## MLflow version

- Client: 3.11.2.dev0 (also reproducible on released builds with workspace support)
- Tracking server: 3.11.2.dev0

## System information

- **OS**: Linux (self-hosted MLflow server)
- **Python version**: 3.10+

## Describe the problem

When MLflow workspaces are enabled and a judge evaluation job is triggered from a **non-default workspace**, the judge runner calls the AI Gateway without forwarding the active workspace in the HTTP headers. As a result, the gateway resolves the endpoint against the **default workspace** instead of the workspace the job belongs to, causing a 404 / endpoint-not-found error for any workspace that doesn't have a duplicate copy of the endpoint in default.

### Root cause

In `mlflow/genai/judges/adapters/gateway_adapter.py`, two code paths build HTTP headers for gateway calls but never inject `X-MLFLOW-WORKSPACE`:

**1. Tool-calling path — `_invoke_and_handle_tools` (line ~522):**
```python
headers = dict(provider_instance.headers or {})
if provider == "gateway":
    headers[MLFLOW_GATEWAY_CALLER_HEADER] = GatewayCaller.JUDGE.value
    # ← X-MLFLOW-WORKSPACE is never added here
```

**2. Non-tool-calling path — `_invoke_via_gateway`:**
```python
# extra_headers passed through as-is; no workspace injection for gateway provider
return score_model_on_payload(..., extra_headers=extra_headers, ...)
```

The workspace **is** available at call time: the job runner correctly sets the `MLFLOW_WORKSPACE` environment variable on the subprocess (see `mlflow/server/jobs/utils.py`, the `MLFLOW_WORKSPACE.name` assignment), and `get_request_workspace()` in `mlflow/utils/workspace_context.py` reads it. It just never gets forwarded as the `X-MLFLOW-WORKSPACE` HTTP header to the gateway.

The constant `WORKSPACE_HEADER_NAME = "X-MLFLOW-WORKSPACE"` is already defined in `mlflow/utils/workspace_utils.py`.

## Code to reproduce issue

```python
import mlflow

# Setup: two workspaces ("default" and "my-workspace"), each with an experiment.
# AI Gateway endpoint "evaluation-endpoint" exists ONLY in "my-workspace".

mlflow.set_workspace("my-workspace")

# Run judge evaluation — this internally calls gateway://evaluation-endpoint
# The job runner sets MLFLOW_WORKSPACE=my-workspace in the subprocess env,
# but gateway_adapter.py never forwards it as X-MLFLOW-WORKSPACE.
results = mlflow.genai.evaluate(
    data=...,
    predict_fn=...,
    scorers=[mlflow.genai.scorer(model="gateway:/evaluation-endpoint", ...)],
)
# ↑ Fails with endpoint-not-found because the gateway receives no workspace header
#   and falls back to resolving "evaluation-endpoint" in the default workspace.
```

## Stack trace

```
MlflowException: Failed to invoke judge model: 404 - Endpoint 'evaluation-endpoint' not found.
  File "mlflow/genai/judges/adapters/gateway_adapter.py", line 605, in _invoke_and_handle_tools
    raise MlflowException(
```

## Proposed fix

**`mlflow/genai/judges/adapters/gateway_adapter.py`**

Add two imports at the top:
```python
from mlflow.utils.workspace_context import get_request_workspace
from mlflow.utils.workspace_utils import WORKSPACE_HEADER_NAME
```

**Fix 1 — `_invoke_and_handle_tools`:**
```python
headers = dict(provider_instance.headers or {})
if provider == "gateway":
    headers[MLFLOW_GATEWAY_CALLER_HEADER] = GatewayCaller.JUDGE.value
    if ws := get_request_workspace():
        headers[WORKSPACE_HEADER_NAME] = ws
```

**Fix 2 — `_invoke_via_gateway`:**
```python
if provider == "gateway":
    workspace_headers = {}
    if ws := get_request_workspace():
        workspace_headers[WORKSPACE_HEADER_NAME] = ws
    extra_headers = {**workspace_headers, **(extra_headers or {})}
```

## What component(s) does this bug affect?

- `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

### LINKED ISSUE #23034: [FR] MCP server exposing run analysis primitives (e.g. `get_metric_history`) for autoresearch
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

No. I cannot contribute this feature at this time.

### Proposal Summary

At [Edison Scientific](https://edisonscientific.com/) we use MLFlow for experiment tracking, and it's great. My request is to expand the MCP server to better accommodate autoresearch-style pipelines. Mainly I desire `get_metric_history` MCP functionality so time series of data can be exported.

### Motivation

> #### What is the use case for this feature?

Letting coding agents use MLFlow as a primary and key tool in hyperparameter optimization of training pipelines.

> #### Why is this use case valuable to support for MLflow users in general?

Autoresearch is the new norm in model training world.

> #### Why is this use case valuable to support for your project(s) or organization?

I can tell Claude Code:

1. We want GRPO running on 4 nodes for [Qwen3-32B](https://huggingface.co/Qwen/Qwen3-32B) with reward increasing across a 1-day training job
2. View MLFlow and run logs to monitor training, adjusting hypers as needed

> #### Why is it currently difficult to achieve this use case?

The MCP server is limited, it doesn't expose capabilities such as:
- Getting a full time series of data
- Searching runs and/or listing artifacts
- Downloading artifacts

### Details

I got Claude Code to tell me what's holding it back, and it's requesting these:

```python
# 1. Per-step metric history (highest priority)
mcp__mlflow.get_metric_history(
    run_id: str,
    metric_key: str,
    max_results: int | None = None,  # optional pagination
) -> list[{"step": int, "value": float, "timestamp": int}]

# Wraps: MlflowClient.get_metric_history(run_id, key)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L478
# REST:  GET /api/2.0/mlflow/metrics/get-history?run_id=...&metric_key=...
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/server/handlers.py#L1820

# 2. Filter runs (v3.11.1 only `list_runs` exists, returning everything)
mcp__mlflow.search_runs(
    experiment_ids: list[str],
    filter_string: str = "",        # MLflow's run-search DSL, e.g. 'metrics.accuracy > 0.9 AND params.lr = "1e-5"'
    order_by: list[str] | None = None,
    max_results: int = 100,
    run_view_type: str = "ACTIVE_ONLY",
) -> list[RunSummary]

# Wraps: MlflowClient.search_runs(...)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L3841
# REST:  POST /api/2.0/mlflow/runs/search
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/server/handlers.py#L1688

# 3. Read run artifacts (configs, eval JSONs, etc.)
mcp__mlflow.list_artifacts(run_id: str, path: str = "") -> list[FileInfo]
mcp__mlflow.download_artifact(run_id: str, path: str) -> bytes  # or a temp file path

# Wraps: MlflowClient.list_artifacts(run_id, path)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L3621
#        and MlflowClient.download_artifacts(...)
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/tracking/client.py#L3680
# REST:  GET /api/2.0/mlflow/artifacts/list
#        https://github.com/mlflow/mlflow/blob/v3.11.1/mlflow/server/handlers.py#L1743

# 4. Convenience: side-by-side comparison
mcp__mlflow.compare_runs(
    run_ids: list[str],
    metric_keys: list[str] | None = None,   # default: union of all metrics
    param_keys: list[str] | None = None,    # default: union of all params
) -> dict  # {"<run_id>": {"metrics": {...}, "params": {...}}}
```

### What machine learning domain(s) is this feature request about?

- [x] `domain/genai`: LLMs, Agents, and other GenAI-related use cases
- [ ] `domain/classical-ml`: Traditional machine learning, such as linear regression.
- [ ] `domain/deep-learning`: Deep learning and neural networks.
- [ ] `domain/platform`: MLflow platform foundation, not specific to a particular machine learning domain.

### What area(s) of MLflow is this feature request about?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

### PR DESCRIPTION
**Pull request by SahilKumar75**
### Related Issues/PRs

Supersedes #23046
Closes #23001
Closes #23034

### What changes are proposed in this pull request?

This PR fixes two issues:

1. **[BUG] Gateway adapter does not forward workspace context when invoking judge endpoints (#23001)**
   - `_invoke_via_gateway`: attaches `X-MLFLOW-WORKSPACE` header when the active workspace is set
   - `_invoke_and_handle_tools`: attaches the same header for the tool-calling loop path
   - Ensures correct endpoint resolution in multi-workspace Databricks deployments

2. **[FR] MCP server exposing run analysis primitives for autoresearch (#23034)**
   - Added `get-metric-history` CLI command (exposed via MCP) for time-series metric data
   - Added `search` command for filtering runs with `--filter-string` / `--order-by`
   - Added `compare` command for side-by-side run comparison
   - Exposed `list` and `download` artifact commands via the `@mlflow_mcp` decorator
   - Registered the `artifacts` tool category in the MCP server

**Additional cleanup (reviewer feedback on #23046):**
- Fixed `ViewType` bug in `mlflow/runs.py`: `view_type` string is now converted to `ViewType` enum before being passed to `client.search_runs`
- Removed "and pagination" from the `search_runs` docstring (the command returns a bounded list, not a paginated cursor)
- Updated the `MLFLOW_MCP_TOOLS` inline comment to explicitly mention artifacts
- Updated `tests/mcp/test_mcp.py::test_list_tools` to include all new artifact tools
- Added unit tests for workspace header forwarding in `tests/genai/judges/adapters/test_gateway_adapter.py`

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

New tests added:
- `tests/genai/judges/adapters/test_gateway_adapter.py`: 4 new tests covering `X-MLFLOW-WORKSPACE` header forwarding for both `_invoke_via_gateway` and `_invoke_and_handle_tools` (gateway provider path and non-gateway provider path)
- `tests/mcp/test_mcp.py`: `test_list_tools` updated to assert all new MCP tools are present

### Does this PR require documentation update?

- [x] No.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.
- [ ] Yes. Please link the corresponding PR or explain how you plan to update it.

#### Is this a user-facing change?

- [ ] No.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

**Bug fix:** Gateway adapter now forwards `X-MLFLOW-WORKSPACE` header when invoking judge endpoints, ensuring correct workspace resolution in multi-workspace Databricks deployments (fixes #23001).

**Feature:** MCP server now exposes run analysis tools (`get_metric_history`, `search_runs`, `compare_runs`) and artifact tools (`list_artifacts`, `download_artifact`) for AI agent autoresearch workflows (fixes #23034).

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [x] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Comment by thkim-us** [comment]
Nice to see the surface expanding. We've been running these same primitives in [`@us-all/mlflow-mcp`](https://github.com/us-all/mlflow-mcp-server) (TypeScript MCP server, 82 tools) for a few weeks — a couple of small gotchas in case they save reviewer cycles:

**`get-metric-history` truncation**: in MLflow 3.x, `/api/2.0/mlflow/metrics/get-history` silently truncates when `max_results` is not passed (only returns `next_page_token`). We had to explicitly pass 25000 (the documented max) to get the full series for runs with many points. Worth confirming the new CLI command does this — autoresearch users often have multi-thousand-point training curves.

**`compare-runs` ergonomics**: when comparing N runs, the LLM consistently asked "which params actually differ?" before doing anything else. Auto-deriving a `differing_params` field in the response saved a round-trip in every workflow. Worth considering if the new `compare` command doesn't already.

**Artifacts via MCP — read flow vs URL flow**: we ship two tools (`list-artifacts`, `get-trace-attachment`) that *return content* directly, plus a `get-model-version-download-uri` for the URL flow. For agents the inline-content path is much higher leverage; for human users the URL is what they want. Both is best.

**For the MCP registry side**: when this lands, would be great to also publish `mlflow[mcp]` to [registry.modelcontextprotocol.io](https://registry.modelcontextprotocol.io/) (we just did `io.github.us-all/mlflow` there using `server.json`). That makes one-click discovery work for MCP-aware clients without users needing to know the package name in advance.

Happy to defer any feature overlap once this lands and stabilizes — keeping ours as a TS alternative for users who can't or don't want to pull in the full Python `mlflow[mcp]` extra. Thanks for the work on this!

**Comment by SahilKumar75** [comment]
I've resolved the merge conflicts with master so this is ready to go! I've also addressed the PR feedback:

1. Fixed the bug with --max-results in the get-metric-history CLI command (negative numbers now throw an error, and 0 correctly returns an empty list).
2. Looked into the truncation concern for get-metric-history. The underlying MlflowClient automatically handles pagination in a while loop (fetching in chunks of 25,000), so the CLI command fetches the full history by default and does not suffer from silent truncation.

**Comment by SahilKumar75** [comment]
DCO passes, all checks pass. Only blocked on maintainer approval. Could someone from the core team take a look? cc @harupy @serena-ruan @B-Step62
